google-cloud-bigquery/src/main/scala/akka/stream/alpakka/googlecloud/bigquery/scaladsl/BigQueryQueries.scala

```diff
@@ -0,0 +1,157 @@
+   * @tparam Out the data model of the query results
+   * @return a [[akka.stream.scaladsl.Source]] that emits an [[Out]] for each row of the results and materializes a [[scala.concurrent.Future]] containing the [[akka.stream.alpakka.googlecloud.bigquery.model.QueryJsonProtocol.QueryResponse]]
+   */
+  def query[Out](query: QueryRequest)(
+      implicit um: FromEntityUnmarshaller[QueryResponse[Out]]
+  ): Source[Out, Future[QueryResponse[Out]]] =
+    Source
+      .fromMaterializer { (mat, attr) =>
+        import BigQueryException._
+        import SprayJsonSupport._
+        import mat.executionContext
+        implicit val system = mat.system
+        implicit val settings = BigQueryAttributes.resolveSettings(attr, mat)
+
+        Source.lazyFutureSource { () =>
+          for {
+            entity <- Marshal(query).to[RequestEntity]
+            initialRequest = HttpRequest(POST, BigQueryEndpoints.queries(settings.projectId), entity = entity)
+            response <- BigQueryHttp().retryRequestWithOAuth(initialRequest)
+            initialQueryResponse <- Unmarshal(response.entity).to[QueryResponse[Out]]
+          } yield {
+
+            val jobId = initialQueryResponse.jobReference.jobId.getOrElse {
+              throw BigQueryException("Query response did not contain job id.")
+            }
+
+            val head =
+              if (initialQueryResponse.jobComplete)
+                Source.single(initialQueryResponse)
+              else
+                Source.empty
+
+            val tail =
+              if (initialQueryResponse.jobComplete & initialQueryResponse.pageToken.isEmpty)
+                Source.empty
+              else {
+                import settings.retrySettings._
+                val pages = queryResultsPages[Out](jobId,
+                                                   None,
+                                                   query.maxResults,
+                                                   query.timeoutMs,
+                                                   initialQueryResponse.jobReference.location,
+                                                   initialQueryResponse.pageToken)
+                  .map(Success(_))
+                  .recover { case ex => Failure(ex) } // Allows upstream failures to escape the RestartSource
+                  .map { queryResponse =>
+                    if (queryResponse.toOption.forall(_.jobComplete))
+                      queryResponse
+                    else
+                      throw BigQueryException("Query job not complete.")
+                  }
+                  .addAttributes(attr)
+                val restartSettings = RestartSettings(minBackoff, maxBackoff, randomFactor)
+                RestartSource.onFailuresWithBackoff(restartSettings)(() => pages).map(_.get)
+              }
+
+            head.concat(tail)
+          }
+
+        }
+      }
+      .wireTapMat(Sink.head)(Keep.right)
```

Comment:
Yes, that's exactly what it does. My initial instinct was to use `alsoTo` but because `Sink.head` cancels after receiving its first element, this would cause the entire stream to cancel. On the other hand `wireTap` is unaffected by the secondary sink (the tap).

Reply:
:+1: